src/bsoncxx/builder/list/document.hpp

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/exception/error_code.hpp>
+#include <bsoncxx/exception/exception.hpp>
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+namespace list {
+using namespace bsoncxx::types;
+
+///
+/// A JSON-like builder for creating BSON documents.
+///
+class document {
+   public:
+    ///
+    /// Default constructor. Creates an empty document.
+    ///
+    document() = default;
+
+    ///
+    /// Creates a document from a key-value pair. Initializes the key with 'key' and the value with
+    /// 'value'. The value may be another document.
+    ///
+    /// @param key
+    ///     the value to initialize the key in the key-value pair.
+    /// @param value
+    ///     the value to initialize the value in the key-value pair.
+    ///
+    /// @note a bsoncxx::types::bson_value::value is direct-initialized from 'value'.
+    ///
+    /// @see bsoncxx::types::bson_value::value.
+    ///
+    template <typename T>
+    document(stdx::string_view key, T&& value) {
+        _core.key_view(key);
+        _core.append(bson_value::value{value});
```

Comment:
since we are taking value by && here, let's `std::move` value into the call on L56

Reply:
The `append` method in `core.hpp` and the copy constructor from `value.hpp` used here take a const reference. Therefore, `value` will be copied regardless. Using `std::move` wouldn't avoid that, I think.

Reply:
Right, the caller can either copy or `std::move` into this method. 
 But regardless of whether the caller sends us that T&& via a move or via copy, it's best practice for us to `std::move` the argument to its final destination (in this case, to the core builder) here.  Using `std::move` to pass `value` here forces the compiler to keep the same value category (as an rvalue reference) for `value` when we pass it into _core.

Reply:
Done... kind of. Forwarding this does the above but also eliminates the need for the other `const T&` constructor too.

Reply:
Oh nice!  Right, these methods are already templated, so std::forward works well.